### PR TITLE
Adds definition and resolved image paths to RtApi.

### PR DESCRIPTION
**Pull request by feldstj**

### CODE REVIEW

File: source/MaterialXRuntime/RtApi.cpp [hunk 1]
```diff
@@ -229,6 +279,43 @@ const FileSearchPath& RtApi::getImplementationSearchPath() const
     return _cast(_ptr)->getImplementationSearchPath();
 }
 
+const FileSearchPath RtApi::getResolvedImagePaths(bool includeSubFolders) const
```
Comment (feldstj) [inline source/MaterialXRuntime/RtApi.cpp:282]: This was pretty taken from LookdevXGraphAPI.

File: source/MaterialXRuntime/Private/PvtApi.h [hunk 2]
```diff
@@ -292,6 +307,70 @@ class PvtApi
...
+        {
+            definitionPaths.append(localDefinitionPaths);
+        }
+        FileSearchPath materialXCoreDefinitionPaths = getMaterialXCoreDefinitionPaths();
+        if (!materialXCoreDefinitionPaths.isEmpty())
+        {
+            definitionPaths.append(materialXCoreDefinitionPaths);
+        }
+
+        if (includeSubFolders)
```
Comment (feldstj) [inline source/MaterialXRuntime/Private/PvtApi.h:354]: Gets the subfolders

Comment (bernardkwok) [reply source/MaterialXRuntime/Private/PvtApi.h:354]: I'm wondering for the order of this whether we should look at core + mtlx definition paths, then local + global as this defines search order.

File: source/MaterialXRuntime/Private/PvtApi.h [hunk 3]
```diff
@@ -292,6 +307,70 @@ class PvtApi
...
+        {
+            FileSearchPath childDefinitionPaths;
+            for (FileSearchPath::Iterator path = definitionPaths.begin(); path != definitionPaths.end(); ++path)
+            {
+                FilePath filePath = *path;
+                FilePathVec subdirs = filePath.getSubDirectories();
+                for (FilePath subdirPath : subdirs)
+                {
+                    childDefinitionPaths.append(subdirPath);
```
Comment (bernardkwok) [inline source/MaterialXRuntime/Private/PvtApi.h:363]: I think we should check for duplicates here. Believe we have the same issue in loadLibraries() so would be good to only append unique paths.

File: source/MaterialXRuntime/RtApi.cpp [hunk 4]
```diff
@@ -229,6 +279,43 @@ const FileSearchPath& RtApi::getImplementationSearchPath() const
...
+        FilePathVec childFolders;
+        getSubdirectories(libraryFolders, searchPath, childFolders);
+        libraryFolders.insert(std::end(libraryFolders), std::begin(childFolders), std::end(childFolders));
+        for (const auto& childFolder : childFolders)
+        {
+            imageSearchPath.append(childFolder);
+        }
+    }
+    // Add working stage Uris to search path
+    for (const FilePath& uri : stage->getSourceUri())
```
Comment (bernardkwok) [inline source/MaterialXRuntime/RtApi.cpp:313]: This seems incorrect. These are paths to file names which are parsed already at 287 to get the parent path.

File: source/MaterialXRuntime/Private/PvtApi.h [hunk 5]
```diff
@@ -292,6 +307,70 @@ class PvtApi
...
+    {
+        return _localDefinitionPaths;
+    }
+
+    const FileSearchPath& getMaterialXCoreDefinitionPaths() const
+    {
+        return _materialXCoreDefinitionPaths;
+    }
+
+    const FileSearchPath getDefinitionPaths(bool includeSubFolders=true) const
```
Comment (bernardkwok) [inline source/MaterialXRuntime/Private/PvtApi.h:330]: I don't see this being used anywhere ?
The location of the definitions is also part of where image should be searched for.

File: source/MaterialXRuntime/RtApi.cpp [hunk 6]
```diff
@@ -229,6 +279,43 @@ const FileSearchPath& RtApi::getImplementationSearchPath() const
...
+    if (includeSubFolders)
+    {
+        FilePathVec childFolders;
+        getSubdirectories(libraryFolders, searchPath, childFolders);
+        libraryFolders.insert(std::end(libraryFolders), std::begin(childFolders), std::end(childFolders));
+        for (const auto& childFolder : childFolders)
+        {
+            imageSearchPath.append(childFolder);
+        }
+    }
```
Comment (bernardkwok) [inline source/MaterialXRuntime/RtApi.cpp:311]: We should add the definition paths here, also if inludeSubFolders is false, nothing seems to be updated.

File: source/MaterialXRuntime/RtApi.cpp [hunk 7]
```diff
@@ -229,6 +279,43 @@ const FileSearchPath& RtApi::getImplementationSearchPath() const
...
+    // Add working stage Uris to search path
+    for (const FilePath& uri : stage->getSourceUri())
+    {
+        imageSearchPath.append(uri);
+    }
+}
```
Comment (bernardkwok) [inline source/MaterialXRuntime/RtApi.cpp:317]: We don't seem to return anything?

File: source/MaterialXRuntime/Private/PvtApi.h (hunk 2 above)
Comment (feldstj) [reply source/MaterialXRuntime/Private/PvtApi.h:354]: Ok, that's easy to change.

File: source/MaterialXRuntime/Private/PvtApi.h (hunk 3 above)
Comment (feldstj) [reply source/MaterialXRuntime/Private/PvtApi.h:363]: Ok, I'll add a set.

File: source/MaterialXRuntime/RtApi.cpp (hunk 4 above)
Comment (feldstj) [reply source/MaterialXRuntime/RtApi.cpp:313]: I pretty much just copy / pasted this code from LookdevXGraphAPI.cpp. Should I just delete this part?

File: source/MaterialXRuntime/Private/PvtApi.h (hunk 5 above)
Comment (feldstj) [reply source/MaterialXRuntime/Private/PvtApi.h:330]: I'm not currently using this anywhere. Should it be used as the library path passed to VnnRuntime::initRuntime?

File: source/MaterialXRuntime/RtApi.cpp (hunk 7 above)
Comment (feldstj) [reply source/MaterialXRuntime/RtApi.cpp:317]: Hmm... good catch, I wonder why it compiled.

File: source/MaterialXRuntime/RtApi.cpp (hunk 6 above)
Comment (feldstj) [reply source/MaterialXRuntime/RtApi.cpp:311]: Ok, so append the definition paths to the imageSearchPath, will do.

File: source/MaterialXRuntime/RtApi.cpp [hunk 8]
```diff
@@ -199,6 +201,54 @@ void RtApi::clearImplementationSearchPath()
     _cast(_ptr)->clearImplementationSearchPath();
 }
 
+/// Get the search path for MaterialX definitions.
+const FileSearchPath& RtApi::getMaterialXDefinitionPaths() const
+{
+    return _cast(_ptr)->getMaterialXCoreDefinitionPaths();
```
Comment (niklasharrysson) [inline source/MaterialXRuntime/RtApi.cpp:207]: Is this one returning the right path? The "core" definitions paths has a separate method already.

File: source/MaterialXRuntime/Private/PvtApi.h [hunk 9]
```diff
@@ -394,6 +478,10 @@ class PvtApi
     FileSearchPath _searchPaths;
     FileSearchPath _implementationSearchPaths;
     FileSearchPath _textureSearchPaths;
+    FileSearchPath _materialXDefinitionPaths;
```
Comment (niklasharrysson) [inline source/MaterialXRuntime/Private/PvtApi.h:481]: Is this one use? Is is never set anywhere as far as I can see?

What is the difference between _materialXDefinitionPaths and _materialXCoreDefinitionPaths?

**Comment by niklasharrysson** [comment]
Out of curiosity, why are all these different definition paths needed? And why are then needed in the RtApi?

Obviously I have not been part of the discussions that preceded this change, but to my eyes it looks a bit like LookdevX application specific concepts that are leaking out into the general runtime API. Are these methods used anywhere inside the API, or only on LookdevX side? If it's the latter it might be more appropriate to keep this as LookdevX specific concepts, in the VnnRuntime class.

This particular change is not a big thing but I just want to keep in mind that MaterialXRuntime will be merged to main MaterialX in the future, and should stay as clean and generally useful as possible. So additions to the API should be carefully considered if they are really needed, or if they can stay on LookdevX side.

**Comment by bernardkwok** [comment]
> Out of curiosity, why are all these different definition paths needed? And why are they needed in the RtApi?
> 
> Obviously I have not been part of the discussions that preceded this change, but to my eyes it looks a bit like LookdevX application specific concepts that are leaking out into the general runtime API. Are these methods used anywhere inside the API, or only on LookdevX side? If it's the latter it might be more appropriate to keep this as LookdevX specific concepts, in the VnnRuntime class.
> 
> This particular change is not a big thing but I just want to keep in mind that MaterialXRuntime will be merged to main MaterialX in the future, and should stay as clean and generally useful as possible. So additions to the API should be carefully considered if they are really needed, or if they can stay on LookdevX side.

Yes, agreed to keep this clean of any "LookdevXness" -- but this is for general asset and library management for which the common "guidelines" we agreed we want to work out. At this point the "user" and "global" pathing concepts not there. Core library root paths + the user core root paths + the materialx default path + the path relative source file mtlx are already in use.

File: source/MaterialXRuntime/Private/PvtApi.h (hunk 9 above)
Comment (feldstj) [reply source/MaterialXRuntime/Private/PvtApi.h:481]: According to Bernard:
materialXCoreDefinitionPaths are: stdlib, pbrlib etc. materialXDefintionPaths come from the Materialx env variable ("MATERIALX_SEARCH_PATH")

Comment (niklasharrysson) [reply source/MaterialXRuntime/Private/PvtApi.h:486]: So then what is the difference between `getMaterialXCoreDefinitionPaths()` and the existing `getSearchPath()`?

Comment (feldstj) [reply source/MaterialXRuntime/Private/PvtApi.h:486]: It seems that the search path points to the application and library paths, not the specific library paths. @bernardkwok may know more.

**Comment by niklasharrysson** [comment]
> Yes, agreed to keep this clean of any "LookdevXness" -- but this is for general asset and library management for which the common "guidelines" we agreed we want to work out. At this point the "user" and "global" pathing concepts not there. Core library root paths + the user core root paths + the materialx default path + the path relative source file mtlx are already in use.

Ok. It just feels like a plethora of different search paths being hard coded as separate API methods, and I wonder if that is what we want in this API.. To my mind the main thing an application needs to tell the API is where to search for definitions, in order to do code generation etc. And this we already have with the setSearchPath method. Where and how an application saves and organizes content feels very application specific, and I just don't see why the application needs to tell the API about that?

**Comment by niklasharrysson** [comment]
It's fine if we have guidelines in the MaterialX spec for how an application may organize its libraries. But not sure why that needs to be represented here, and especially in hard coded form, since guidelines may change.

If there is a good reason for having it represented here I wonder if it would be better to have an interface like this, instead of hard coding the asset types into separate methods:
```
void setSearchPath(const FileSearchPath& path, const RtIdentifier& assetType = EMPTY_IDENTIFIER);
const FileSearchPath& getSearchPath(const RtIdentifier& assetType = EMPTY_IDENTIFIER);
```
So internally it's a map of search paths for different asset types, and an application may customize it's use in different ways.

But again, I'm not convinced there is a need for having this represented here at all. Sorry for being grumpy, I might very well be missing something.. :)

**Comment by bernardkwok** [comment]
> It's fine if we have guidelines in the MaterialX spec for how an application may organize its libraries. But not sure why that needs to be represented here, and especially in hard coded form, since guidelines may change.
> 
> If there is a good reason for having it represented here I wonder if it would be better to have an interface like this, instead of hard coding the asset types into separate methods:
> 
> ```
> void setSearchPath(const FileSearchPath& path, const RtIdentifier& assetType = EMPTY_IDENTIFIER);
> const FileSearchPath& getSearchPath(const RtIdentifier& assetType = EMPTY_IDENTIFIER);
> ```
> 
> So internally it's a map of search paths for different asset types, and an application may customize it's use in different ways.
> 
> But again, I'm not convinced there is a need for having this represented here at all. Sorry for being grumpy, I might very well be missing something.. :)

Each item except for the core path (stdlib etc) is a user settable item and each can be set independently -- vs having a single search path as is currently the case. The final paths for definitions and images are composed as needed based on changes to each of the individual items which can occur dynamically. These can be settable from command line (if exposed on an application) or as environment variables. LookdevX is basically an editor which can have UI to set the environment variables, but the search paths aren't tied to it.

File: source/MaterialXRuntime/Private/PvtApi.h (hunk 9 above)
Comment (bernardkwok) [reply source/MaterialXRuntime/Private/PvtApi.h:486]: We may be able to get away with not have materialXDefinitionPaths() as the API exists in MaterialX core already. I'm not sure if we have this access in all places. We could leave it out for now if the current context includes materialx access (which is true for the export case).

**Review by bernardkwok — APPROVED** [review]
Looks good to me. @niklasharrysson we should resolve your concerns first.

**Comment by bernardkwok** [comment]
PR no longer needed. Closing.

**Pull request closed without merging**